### PR TITLE
feat: expose not contains key operator

### PR DESCRIPTION
**Pull request by aaron-steinfeld**
## Description

Exposing existing query service operator NOT_CONTAINS_KEY

### CODE REVIEW

**Review by surajpuvvada — DISMISSED** [review]

**Review by tim-mwangi — DISMISSED** [review]

**Review by aaron-steinfeld — COMMENTED** [review]
these changes creeped out more than I meant them too... I can separate if desired. Either way, will separate the CI changes for the _rest_ of this chain of PRs.

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/AttributeCacheKey.java [hunk 1]
```diff
@@ -3,6 +3,9 @@
 import java.util.Map;
 import java.util.Objects;
 
+// TODO replace with org.hypertrace.core.grpcutils.context.ContextualKey
```
Comment (aaron-steinfeld) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/AttributeCacheKey.java:6]: Stopped the scope creep here, but request context has a generated cache key that was added several years ago and can replace this.

File: gateway-service-impl/src/test/java/org/hypertrace/gateway/service/common/AttributeCacheKeyTest.java [hunk 2]
```diff
@@ -48,7 +49,8 @@ public void testAttributeCacheKey() {
   private RequestContext buildContext(String tenantId, Map<String, String> headers) {
     org.hypertrace.core.grpcutils.context.RequestContext grpcContext =
         org.hypertrace.core.grpcutils.context.RequestContext.forTenantId(tenantId);
-    headers.forEach(grpcContext::add);
+    grpcContext.removeHeader(RequestContextConstants.REQUEST_ID_HEADER_KEY);
```
Comment (aaron-steinfeld) [inline gateway-service-impl/src/test/java/org/hypertrace/gateway/service/common/AttributeCacheKeyTest.java:52]: grpc context generate a unique rest id from a while back - removing it for the purpose of this test since comparisons are done across headers.

File: gateway-service-impl/src/test/java/org/hypertrace/gateway/service/common/datafetcher/EntityInteractionsFetcherTest.java [hunk 3]
```diff
@@ -180,8 +180,7 @@ public void testFetchInteractionsIds_withInteractionFilter() {
     EntityInteractionsFetcher interactionsFetcher =
         new EntityInteractionsFetcher(queryServiceClient, attributeMetadataProvider, queryExecutor);
     List<EntityKey> entityKeys =
-        interactionsFetcher.fetchInteractionsIds(
-            new RequestContext(forTenantId(TENANT_ID)), request);
+        interactionsFetcher.fetchInteractionsIds(entitiesRequestContext, request);
```
Comment (aaron-steinfeld) [inline gateway-service-impl/src/test/java/org/hypertrace/gateway/service/common/datafetcher/EntityInteractionsFetcherTest.java:183]: same issue as above - the mock and call both construct their own request ID with differing request ids, and an eq() matcher is used on line 174. Using the same request context hits the matcher as it used to.

**Review by avinashkolluru — APPROVED** [review]

**Comment by aaron-steinfeld** [comment]
no-op amend on last commit to retrigger due to GHA outage earlier.

File: .github/workflows/pr-build.yml [hunk 4]
```diff
@@ -42,44 +30,20 @@ jobs:
...
           args: build dockerBuildImages
 
-      - name: Determine docker tag
-        id: tag
-        run: echo ::set-output name=tag::$(./gradlew -q printDockerImageDefaultTag | head -1)
-
-      - name: Scan docker image
-        uses: azure/container-scan@v0.1
+      - name: Run Trivy vulnerability scanner
+        uses: hypertrace/github-actions/trivy-image-scan@main
```
[the remark this answers is not in the document]
Comment (ravisingal) [reply .github/workflows/pr-build.yml:34]: https://github.com/azure/container-scan repo is now archived, will switch to trivy in all HT repos.

**Review by tim-mwangi — APPROVED** [review]

**Comment by avinashkolluru** [comment]
why does the snyk-scan show up even after removing it?

**Comment by aaron-steinfeld** [comment]
`pull_request_target` workflows use the CI config from the base branch. It's a security feature for public repos, so untrusted PRs from forks can still run CI with secrets, but can't change the CI config to access them. `pull_request` on the other hand can't access secrets on forks (but runs with the CI config from the PR branch).

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target

So snyk scan ran here because it was pull_request_target-triggered, but after merge it stopped running on main and future PRs.
